File: dotnet/src/demos/dataFlowAnalyzer/dataflow/model/ModelBindingManager.cs
using System.Collections;
using System.Collections.Generic;

namespace gudusoft.gsqlparser.demos.dlineage.dataflow.model
{

    using ESetOperatorType = gudusoft.gsqlparser.ESetOperatorType;
    using TMergeInsertClause = gudusoft.gsqlparser.nodes.TMergeInsertClause;
    using TMergeUpdateClause = gudusoft.gsqlparser.nodes.TMergeUpdateClause;
    using TObjectName = gudusoft.gsqlparser.nodes.TObjectName;
    using TObjectNameList = gudusoft.gsqlparser.nodes.TObjectNameList;
    using TParseTreeNode = gudusoft.gsqlparser.nodes.TParseTreeNode;
    using TResultColumnList = gudusoft.gsqlparser.nodes.TResultColumnList;
    using TTable = gudusoft.gsqlparser.nodes.TTable;
    using TCreateViewSqlStatement = gudusoft.gsqlparser.stmt.TCreateViewSqlStatement;
    using TSelectSqlStatement = gudusoft.gsqlparser.stmt.TSelectSqlStatement;
    using TUpdateSqlStatement = gudusoft.gsqlparser.stmt.TUpdateSqlStatement;
    using System;
    using gudusoft.gsqlparser.nodes;
    using gudusoft.gsqlparser;

    public class ModelBindingManager
    {

        private static readonly IDictionary modelBindingMap = new Hashtable();
        private static readonly IDictionary viewModelBindingMap = new Hashtable();
        private static readonly IDictionary insertModelBindingMap = new Hashtable();
        private static readonly IDictionary createModelBindingMap = new Hashtable();
        private static readonly IDictionary mergeModelBindingMap = new Hashtable();
        private static readonly IDictionary updateModelBindingMap = new Hashtable();
        private static readonly List<Relation> relationHolder = new List<Relation>();

        public static void bindModel(object gspModel, object relationModel)
        {
            modelBindingMap[gspModel] = relationModel;
        }

        public static object getModel(object gspModel)
        {
            if (gspModel == null)
                return null;

            if (gspModel is TTable)
            {
                TTable table = (TTable)gspModel;
                if (table.CTE != null)
                {
                    return modelBindingMap[table.CTE];
                }
                if (table.Subquery != null && table.Subquery.ResultColumnList != null)
                {
                    return modelBindingMap[table.Subquery.ResultColumnList];
                }
            }

            if (gspModel is TSelectSqlStatement)
            {
                TSelectSqlStatement select = (TSelectSqlStatement)gspModel;
                if (select.ResultColumnList != null)
                {
                    return modelBindingMap[select.ResultColumnList];
                }
            }

            object result = modelBindingMap[gspModel];
            if (result == null)
            {
                result = createModelBindingMap[gspModel];
            }
            if (result == null)
            {
                result = insertModelBindingMap[gspModel];
            }
            if (result == null)
            {
                result = updateModelBindingMap[gspModel];
            }
            if (result == null)
            {
                result = mergeModelBindingMap[gspModel];
            }
            if (result == null)
            {
                result = viewModelBindingMap[gspModel];
            }
            return result;
        }

        public static void bindViewModel(object gspModel, object relationModel)
        {
            viewModelBindingMap[gspModel] = relationModel;
        }

        public static object getViewModel(object gspModel)
        {
            return viewModelBindingMap[gspModel];
        }

        public static void bindUpdateModel(object gspModel, object relationModel)
        {
            updateModelBindingMap[gspModel] = relationModel;
        }

        public static object getUpdateModel(object gspModel)
        {
            return updateModelBindingMap[gspModel];
        }

        public static void bindMergeModel(object gspModel, object relationModel)
        {
            mergeModelBindingMap[gspModel] = relationModel;
        }

        public static object getMergeModel(object gspModel)
        {
            return mergeModelBindingMap[gspModel];
        }

        public static void bindInsertModel(object gspModel, object relationModel)
        {
            insertModelBindingMap[gspModel] = relationModel;
        }

        public static object getInsertModel(object gspModel)
        {
            return insertModelBindingMap[gspModel];
        }

        public static Table getCreateTable(TTable table)
        {
            if (table != null && table.TableName != null)
            {
                IEnumerator iter = createModelBindingMap.Keys.GetEnumerator();
                while (iter.MoveNext())
                {
                    TTable node = (TTable)iter.Current;
                    if (node.FullName.Equals(table.FullName))
                    {
                        return (Table)createModelBindingMap[node];
                    }
                }
            }
            return null;
        }

        public static Table getCreateModel(TTable table)
        {
            return (Table)createModelBindingMap[table];
        }

        public static void bindCreateModel(TTable table, Table tableModel)
        {
            createModelBindingMap[table] = tableModel;
        }

        public static TTable getTable(TCustomSqlStatement stmt, TObjectName column)
        {
            IEnumerator iter = modelBindingMap.Values.GetEnumerator();
            while (iter.MoveNext())
            {
                object key = iter.Current;
                if (key is Table)
                {
                    key = ((Table)key).TableObject;
                }
                else if (key is QueryTable)
                {
                    key = ((QueryTable)key).TableObject;
                }
                else
                {
                    continue;
                }

                TTable table = (TTable)key;

                if (table.Subquery == stmt)
                    continue;

                if (column.TableString != null && column.TableString.Trim().Length > 0)
                {
                    if (table.AliasName != null)
                    {
                        if (!table.AliasName.Equals(column.TableString))
                        {
                            continue;
                        }
                        else
                        {
                            return table;
                        }
                    }
                }

                TObjectName[] columns = getTableColumns(table);
                for (int i = 0; i < columns.Length; i++)
                {
                    TObjectName columnName = columns[i];
                    if ("*".Equals(columnName.ColumnNameOnly))
                    {
                        continue;
                    }
                    if (columnName == column)
                    {
                        return table;
                    }
                }
            }
            return null;
        }

        public static TObjectName[] getTableColumns(TTable table)
        {
            Table createTable = ModelBindingManager.getCreateTable(table);
            if (createTable != null)
            {
                IList<TableColumn> columnList = createTable.Columns;
                TObjectName[] columns = new TObjectName[columnList.Count];
                for (int i = 0; i < columns.Length; i++)
                {
                    columns[i] = columnList[i].ColumnObject;
                }
                Array.Sort(columns, new ComparatorAnonymousInnerClass());
                return columns;
            }
            else
            {
                TObjectNameList list = table.ObjectNameReferences;
                List<TObjectName> columns = new List<TObjectName>();

                if (list.size() == 0 && table.Subquery != null)
                {
                    ResultSet resultSet = (ResultSet)ModelBindingManager.getModel(table.Subquery);
                    if (resultSet != null)
                    {
                        IList<ResultColumn> columnList = resultSet.Columns;
                        for (int i = 0; i < columnList.Count; i++)
                        {
                            ResultColumn resultColumn = columnList[i];
                            if (resultColumn.ColumnObject is TResultColumn)
                            {
                                TResultColumn columnObject = ((TResultColumn)resultColumn.ColumnObject);
                                TAliasClause alias = columnObject.AliasClause;
                                if (alias != null && alias.AliasName != null)
                                {
                                    columns.Add(alias.AliasName);
                                }
                                else
                                {
                                    if (columnObject.FieldAttr != null)
                                    {
                                        columns.Add(columnObject.FieldAttr);
                                    }
                                    else
                                    {
                                        continue;
                                    }
                                }
                            }
                            else if (resultColumn.ColumnObject is TObjectName)
                            {
                                columns.Add((TObjectName)resultColumn.ColumnObject);
                            }
                        }
                    }
                }
                else
                {
                    for (int i = 0; i < list.size(); i++)
                    {
                        columns.Add(list.getObjectName(i));
                    }
                }

                TObjectName[] columnArray = columns.ToArray();
                Array.Sort(columnArray, new ComparatorAnonymousInnerClass());
                return columnArray;
            }

        }

        private class ComparatorAnonymousInnerClass : IComparer<TObjectName>
        {
            public virtual int Compare(TObjectName o1, TObjectName o2)
            {
                return o1.startToken.posinlist - o2.startToken.posinlist;
            }
        }

        public static IList<TTable> BaseTables
        {
            get
            {
                IList<TTable> tables = new List<TTable>();

                IEnumerator iter = modelBindingMap.Keys.GetEnumerator();
                while (iter.MoveNext())
                {
                    object key = iter.Current;
                    if (!(key is TTable))
                    {
                        continue;
                    }
                    TTable table = (TTable)key;
                    if (table.Subquery == null)
                    {
                        tables.Add(table);
                    }
                }

                iter = createModelBindingMap.Keys.GetEnumerator();
                while (iter.MoveNext())
                {
                    object key = iter.Current;
                    if (!(key is TTable))
                    {
                        continue;
                    }
                    TTable table = (TTable)key;
                    tables.Add(table);
                }

                iter = insertModelBindingMap.Keys.GetEnumerator();
                while (iter.MoveNext())
                {
                    object key = iter.Current;
                    if (!(key is TTable))
                    {
                        continue;
                    }
                    TTable table = (TTable)key;
                    tables.Add(table);
                }

                iter = mergeModelBindingMap.Keys.GetEnumerator();
                while (iter.MoveNext())
                {
                    object key = iter.Current;
                    if (!(key is TTable))
                    {
                        continue;
                    }
                    TTable table = (TTable)key;
                    tables.Add(table);
                }

                iter = updateModelBindingMap.Keys.GetEnumerator();
                while (iter.MoveNext())
                {
                    object key = iter.Current;
                    if (!(key is TTable))
                    {
                        continue;
                    }
                    TTable table = (TTable)key;
                    tables.Add(table);
                }

                return tables;
            }
        }

        public static IList<TCreateViewSqlStatement> Views
        {
            get
            {
                IList<TCreateViewSqlStatement> views = new List<TCreateViewSqlStatement>();

                IEnumerator iter = viewModelBindingMap.Keys.GetEnumerator();
                while (iter.MoveNext())
                {
                    object key = iter.Current;
                    if (!(key is TCreateViewSqlStatement))
                    {
                        continue;
                    }
                    TCreateViewSqlStatement view = (TCreateViewSqlStatement)key;
                    views.Add(view);
                }
                return views;
            }
        }

        public static IList<TResultColumnList> SelectResultSets
        {
            get
            {
                IList<TResultColumnList> resultSets = new List<TResultColumnList>();

                IEnumerator iter = modelBindingMap.Keys.GetEnumerator();
                while (iter.MoveNext())
                {
                    object key = iter.Current;
                    if (!(key is TResultColumnList))
                    {
                        continue;
                    }
                    TResultColumnList resultset = (TResultColumnList)key;
                    resultSets.Add(resultset);
                }
                return resultSets;
            }
        }

        public static IList<TSelectSqlStatement> SelectSetResultSets
        {
            get
            {
                IList<TSelectSqlStatement> resultSets = new List<TSelectSqlStatement>();

                IEnumerator iter = modelBindingMap.Keys.GetEnumerator();
                while (iter.MoveNext())
                {
                    object key = iter.Current;
                    if (!(key is TSelectSqlStatement))
                    {
                        continue;
                    }

                    TSelectSqlStatement stmt = (TSelectSqlStatement)key;
                    if (stmt.SetOperatorType == ESetOperatorType.none)
                    {
                        continue;
                    }

                    resultSets.Add(stmt);
                }
                return resultSets;
            }
        }

        public static IList<TCTE> CTEs
        {
            get
            {
                IList<TCTE> ctes = new List<TCTE>();

                IEnumerator iter = modelBindingMap.Keys.GetEnumerator();
                while (iter.MoveNext())
                {
                    object key = iter.Current;
                    if (!(key is TCTE))
                    {
                        continue;
                    }

                    TCTE cte = (TCTE)key;
                    ctes.Add(cte);
                }
                return ctes;
            }
        }

        public static IList<TTable> TableWithSelectSetResultSets
        {
            get
            {
                IList<TTable> resultSets = new List<TTable>();

                IEnumerator iter = modelBindingMap.Keys.GetEnumerator();
                while (iter.MoveNext())
                {
                    object key = iter.Current;
                    if (!(key is TTable))
                    {
                        continue;
                    }

                    if (((TTable)key).Subquery == null)
                    {
                        continue;
                    }
                    TSelectSqlStatement stmt = ((TTable)key).Subquery;
                    if (stmt.SetOperatorType == ESetOperatorType.none)
                    {
                        continue;
                    }

                    resultSets.Add((TTable)key);
                }
                return resultSets;
            }
        }

        public static IList<TParseTreeNode> MergeResultSets
        {
            get
            {
                IList<TParseTreeNode> resultSets = new List<TParseTreeNode>();

                IEnumerator iter = modelBindingMap.Keys.GetEnumerator();
                while (iter.MoveNext())
                {
                    object key = iter.Current;
                    if (!(key is TMergeUpdateClause) && !(key is TMergeInsertClause))
                    {
                        continue;
                    }
                    TParseTreeNode resultset = (TParseTreeNode)key;
                    resultSets.Add(resultset);
                }
                return resultSets;
            }
        }

        public static IList<TParseTreeNode> UpdateResultSets
        {
            get
            {
                IList<TParseTreeNode> resultSets = new List<TParseTreeNode>();

                IEnumerator iter = modelBindingMap.Keys.GetEnumerator();
                while (iter.MoveNext())
                {
                    object key = iter.Current;
                    if (!(key is TUpdateSqlStatement))
                    {
                        continue;
                    }
                    TParseTreeNode resultset = (TParseTreeNode)key;
                    resultSets.Add(resultset);
                }
                return resultSets;
            }
        }

        public static void addRelation(Relation relation)
        {
            if (relation != null && !relationHolder.Contains(relation))
            {
                relationHolder.Add(relation);
            }
        }

        public static Relation[] Relations
        {
            get
            {
                return relationHolder.ToArray();
            }
        }

        public static void reset()
        {
            relationHolder.Clear();
            modelBindingMap.Clear();
            viewModelBindingMap.Clear();
            insertModelBindingMap.Clear();
            mergeModelBindingMap.Clear();
            updateModelBindingMap.Clear();
            createModelBindingMap.Clear();
        }

    }

}

File: java/src/demos/dlineage/dataflow/model/ModelBindingManager.java

package demos.dlineage.dataflow.model;

import gudusoft.gsqlparser.ESetOperatorType;
import gudusoft.gsqlparser.TCustomSqlStatement;
import gudusoft.gsqlparser.nodes.TAliasClause;
import gudusoft.gsqlparser.nodes.TCTE;
import gudusoft.gsqlparser.nodes.TMergeInsertClause;
import gudusoft.gsqlparser.nodes.TMergeUpdateClause;
import gudusoft.gsqlparser.nodes.TObjectName;
import gudusoft.gsqlparser.nodes.TObjectNameList;
import gudusoft.gsqlparser.nodes.TParseTreeNode;
import gudusoft.gsqlparser.nodes.TResultColumn;
import gudusoft.gsqlparser.nodes.TResultColumnList;
import gudusoft.gsqlparser.nodes.TTable;
import gudusoft.gsqlparser.nodes.TTableList;
import gudusoft.gsqlparser.stmt.TCreateViewSqlStatement;
import gudusoft.gsqlparser.stmt.TCursorDeclStmt;
import gudusoft.gsqlparser.stmt.TSelectSqlStatement;
import gudusoft.gsqlparser.stmt.TUpdateSqlStatement;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.CopyOnWriteArrayList;

import demos.dlineage.util.Pair;

@SuppressWarnings({
		"unchecked", "rawtypes"
})
public class ModelBindingManager
{

	private static final Map modelBindingMap = new LinkedHashMap( );
	private static final Map viewModelBindingMap = new LinkedHashMap( );
	private static final Map insertModelBindingMap = new LinkedHashMap( );
	private static final Map createModelBindingMap = new LinkedHashMap( );
	private static final Map createModelQuickBindingMap = new HashMap( );
	private static final Map mergeModelBindingMap = new LinkedHashMap( );
	private static final Map updateModelBindingMap = new LinkedHashMap( );
	private static final Map cursorModelBindingMap = new LinkedHashMap( );
	private static final List<Relation> relationHolder = new CopyOnWriteArrayList<Relation>( );

	private static final Map<String, TTable> tableAliasMap = new HashMap( );
	private static final Set tableSet = new HashSet( );

	public static void bindModel( Object gspModel, Object relationModel )
	{
		modelBindingMap.put( gspModel, relationModel );

		TTable table = null;
		if ( gspModel instanceof TTable )
		{
			table = ( (TTable) gspModel );
		}
		else if ( gspModel instanceof Table )
		{
			table = ( (Table) gspModel ).getTableObject( );
		}
		else if ( gspModel instanceof QueryTable )
		{
			table = ( (QueryTable) gspModel ).getTableObject( );
		}
		else if ( gspModel instanceof TCTE )
		{
			TTableList tables = ( (TCTE) gspModel ).getPreparableStmt( ).tables;
			for ( int j = 0; j < tables.size( ); j++ )
			{
				TTable item = tables.getTable( j );
				if ( item != null && item.getAliasName( ) != null )
				{
					tableAliasMap.put( item.getAliasName( ).toLowerCase( ),
							item );
				}

				if ( item != null )
				{
					tableSet.add( item );
				}
			}
		}
		else if ( gspModel instanceof Pair
				&& ( (Pair) gspModel ).first instanceof Table )
		{
			table = ( (Table) ( (Pair) gspModel ).first ).getTableObject( );
		}
		
		if ( table == null && relationModel instanceof QueryTable )
		{
			table = ( (QueryTable) relationModel ).getTableObject( );
		}

		if ( table != null && table.getAliasName( ) != null )
		{
			tableAliasMap.put( table.getAliasName( ).toLowerCase( ), table );
		}

		if ( table != null )
		{
			tableSet.add( table );
		}
	}

	public static Object getModel( Object gspModel )
	{
		if ( gspModel == null )
		{
			return null;
		}

		if ( gspModel instanceof TTable )
		{
			TTable table = (TTable) gspModel;
			if ( table.getCTE( ) != null )
			{
				return modelBindingMap.get( table.getCTE( ) );
			}
			if ( table.getSubquery( ) != null
					&& table.getSubquery( ).getResultColumnList( ) != null )
			{
				return modelBindingMap
						.get( table.getSubquery( ).getResultColumnList( ) );
			}
		}
		if ( gspModel instanceof TSelectSqlStatement )
		{
			TSelectSqlStatement select = (TSelectSqlStatement) gspModel;
			if ( select.getResultColumnList( ) != null )
			{
				return modelBindingMap.get( select.getResultColumnList( ) );
			}
		}
		Object result = modelBindingMap.get( gspModel );
		if ( result == null )
		{
			result = createModelBindingMap.get( gspModel );
		}
		if ( result == null )
		{
			result = insertModelBindingMap.get( gspModel );
		}
		if ( result == null )
		{
			result = updateModelBindingMap.get( gspModel );
		}
		if ( result == null )
		{
			result = mergeModelBindingMap.get( gspModel );
		}
		if ( result == null )
		{
			result = viewModelBindingMap.get( gspModel );
		}
		if ( result == null )
		{
			result = cursorModelBindingMap.get( gspModel );
		}
		return result;
	}

	public static void bindViewModel( Object gspModel, Object relationModel )
	{
		viewModelBindingMap.put( gspModel, relationModel );
	}

	public static Object getViewModel( Object gspModel )
	{
		return viewModelBindingMap.get( gspModel );
	}

	public static void bindUpdateModel( Object gspModel, Object relationModel )
	{
		updateModelBindingMap.put( gspModel, relationModel );
	}

	public static Object getUpdateModel( Object gspModel )
	{
		return updateModelBindingMap.get( gspModel );
	}

	public static void bindMergeModel( Object gspModel, Object relationModel )
	{
		mergeModelBindingMap.put( gspModel, relationModel );
	}

	public static Object getMergeModel( Object gspModel )
	{
		return mergeModelBindingMap.get( gspModel );
	}

	public static void bindInsertModel( Object gspModel, Object relationModel )
	{
		insertModelBindingMap.put( gspModel, relationModel );
	}

	public static Object getInsertModel( Object gspModel )
	{
		return insertModelBindingMap.get( gspModel );
	}

	public static Table getCreateTable( TTable table )
	{
		if ( table != null && table.getTableName( ) != null )
		{
			// Iterator iter = createModelBindingMap.keySet( ).iterator( );
			// while ( iter.hasNext( ) )
			// {
			// TTable node = (TTable) iter.next( );
			// if ( node.getFullName( ).equals( table.getFullName( ) ) )
			// {
			// return (Table) createModelBindingMap.get( node );
			// }
			// }
			return (Table) createModelQuickBindingMap
					.get( table.getFullName( ) );
		}
		return null;
	}

	public static Table getCreateModel( TTable table )
	{
		return (Table) createModelBindingMap.get( table );
	}

	public static void bindCreateModel( TTable table, Table tableModel )
	{
		createModelBindingMap.put( table, tableModel );
		if ( !createModelQuickBindingMap.containsKey( table.getFullName( ) ) )
		{
			createModelQuickBindingMap.put( table.getFullName( ), tableModel );
		}
	}

	public static TObjectName[] getTableColumns( TTable table )
	{
		Table createTable = ModelBindingManager.getCreateTable( table );
		if ( createTable != null )
		{
			List<TableColumn> columnList = createTable.getColumns( );
			TObjectName[] columns = new TObjectName[columnList.size( )];
			for ( int i = 0; i < columns.length; i++ )
			{
				columns[i] = columnList.get( i ).getColumnObject( );
			}
			Arrays.sort( columns, new Comparator<TObjectName>( ) {

				@Override
				public int compare( TObjectName o1, TObjectName o2 )
				{
					return o1.getStartToken( ).posinlist
							- o2.getStartToken( ).posinlist;
				}
			} );
			return columns;
		}

		TObjectNameList list = table.getObjectNameReferences( );
		List<TObjectName> columns = new ArrayList<TObjectName>( );

		if ( table.getCTE( ) != null )
		{
			ResultSet resultSet = (ResultSet) ModelBindingManager
					.getModel( table.getCTE( ) );
			if ( resultSet != null )
			{
				List<ResultColumn> columnList = resultSet.getColumns( );
				for ( int i = 0; i < columnList.size( ); i++ )
				{
					ResultColumn resultColumn = columnList.get( i );
					if ( resultColumn
							.getColumnObject( ) instanceof TResultColumn )
					{
						TResultColumn columnObject = ( (TResultColumn) resultColumn
								.getColumnObject( ) );
						TAliasClause alias = columnObject.getAliasClause( );
						if ( alias != null && alias.getAliasName( ) != null )
						{
							columns.add( alias.getAliasName( ) );
						}
						else
						{
							if ( columnObject.getFieldAttr( ) != null )
							{
								columns.add( columnObject.getFieldAttr( ) );
							}
							else
							{
								continue;
							}
						}
					}
					else if ( resultColumn
							.getColumnObject( ) instanceof TObjectName )
					{
						columns.add(
								(TObjectName) resultColumn.getColumnObject( ) );
					}
				}
			}
		}
		else if ( list.size( ) == 0 && table.getSubquery( ) != null )
		{
			ResultSet resultSet = (ResultSet) ModelBindingManager
					.getModel( table.getSubquery( ) );
			if ( resultSet != null )
			{
				List<ResultColumn> columnList = resultSet.getColumns( );
				for ( int i = 0; i < columnList.size( ); i++ )
				{
					ResultColumn resultColumn = columnList.get( i );
					if ( resultColumn
							.getColumnObject( ) instanceof TResultColumn )
					{
						TResultColumn columnObject = ( (TResultColumn) resultColumn
								.getColumnObject( ) );
						TAliasClause alias = columnObject.getAliasClause( );
						if ( alias != null && alias.getAliasName( ) != null )
						{
							columns.add( alias.getAliasName( ) );
						}
						else
						{
							if ( columnObject.getFieldAttr( ) != null )
							{
								columns.add( columnObject.getFieldAttr( ) );
							}
							else
							{
								continue;
							}
						}
					}
					else if ( resultColumn
							.getColumnObject( ) instanceof TObjectName )
					{
						columns.add(
								(TObjectName) resultColumn.getColumnObject( ) );
					}
				}
			}
		}
		else
		{
			for ( int i = 0; i < list.size( ); i++ )
			{
				columns.add( list.getObjectName( i ) );
			}
		}
		Collections.sort( columns, new Comparator<TObjectName>( ) {

			@Override
			public int compare( TObjectName o1, TObjectName o2 )
			{
				return o1.getStartToken( ).posinlist
						- o2.getStartToken( ).posinlist;
			}
		} );
		return columns.toArray( new TObjectName[0] );
	}

	public static TTable getTable( TCustomSqlStatement stmt,
			TObjectName column )
	{
		if ( column.getTableString( ) != null
				&& column.getTableString( ).trim( ).length( ) > 0 )
		{
			TTable table = tableAliasMap
					.get( column.getTableString( ).toLowerCase( ) );

			if ( table != null && table.getSubquery( ) != stmt )
				return table;
		}

		Iterator iter = tableSet.iterator( );
		while ( iter.hasNext( ) )
		{
			TTable table = (TTable) iter.next( );

			if ( table.getSubquery( ) == stmt )
				continue;

			TObjectName[] columns = getTableColumns( table );
			for ( int i = 0; i < columns.length; i++ )
			{
				TObjectName columnName = columns[i];
				if ( "*".equals( columnName.getColumnNameOnly( ) ) )
					continue;
				if ( columnName == column )
				{
					if ( columnName.getSourceTable( ) == null
							|| columnName.getSourceTable( ) == table )
					{
						return table;
					}
				}
			}
		}
		return null;
	}

	public static List<TTable> getBaseTables( )
	{
		List<TTable> tables = new ArrayList<TTable>( );

		Iterator iter = modelBindingMap.keySet( ).iterator( );
		while ( iter.hasNext( ) )
		{
			Object key = iter.next( );
			if ( !( key instanceof TTable ) )
			{
				continue;
			}
			TTable table = (TTable) key;
			if ( table.getSubquery( ) == null )
			{
				tables.add( table );
			}
		}

		iter = createModelBindingMap.keySet( ).iterator( );
		while ( iter.hasNext( ) )
		{
			Object key = iter.next( );
			if ( !( key instanceof TTable ) )
			{
				continue;
			}
			TTable table = (TTable) key;
			tables.add( table );
		}

		iter = insertModelBindingMap.keySet( ).iterator( );
		while ( iter.hasNext( ) )
		{
			Object key = iter.next( );
			if ( !( key instanceof TTable ) )
			{
				continue;
			}
			TTable table = (TTable) key;
			tables.add( table );
		}

		iter = mergeModelBindingMap.keySet( ).iterator( );
		while ( iter.hasNext( ) )
		{
			Object key = iter.next( );
			if ( !( key instanceof TTable ) )
			{
				continue;
			}
			TTable table = (TTable) key;
			tables.add( table );
		}

		iter = updateModelBindingMap.keySet( ).iterator( );
		while ( iter.hasNext( ) )
		{
			Object key = iter.next( );
			if ( !( key instanceof TTable ) )
			{
				continue;
			}
			TTable table = (TTable) key;
			tables.add( table );
		}

		return tables;
	}

	public static List<TCreateViewSqlStatement> getViews( )
	{
		List<TCreateViewSqlStatement> views = new ArrayList<TCreateViewSqlStatement>( );

		Iterator iter = viewModelBindingMap.keySet( ).iterator( );
		while ( iter.hasNext( ) )
		{
			Object key = iter.next( );
			if ( !( key instanceof TCreateViewSqlStatement ) )
			{
				continue;
			}
			TCreateViewSqlStatement view = (TCreateViewSqlStatement) key;
			views.add( view );
		}
		return views;
	}

	public static List<TResultColumnList> getSelectResultSets( )
	{
		List<TResultColumnList> resultSets = new ArrayList<TResultColumnList>( );

		Iterator iter = modelBindingMap.keySet( ).iterator( );
		while ( iter.hasNext( ) )
		{
			Object key = iter.next( );
			if ( !( key instanceof TResultColumnList ) )
			{
				continue;
			}
			TResultColumnList resultset = (TResultColumnList) key;
			resultSets.add( resultset );
		}
		return resultSets;
	}

	public static List<TSelectSqlStatement> getSelectSetResultSets( )
	{
		List<TSelectSqlStatement> resultSets = new ArrayList<TSelectSqlStatement>( );

		Iterator iter = modelBindingMap.keySet( ).iterator( );
		while ( iter.hasNext( ) )
		{
			Object key = iter.next( );
			if ( !( key instanceof TSelectSqlStatement ) )
			{
				continue;
			}

			TSelectSqlStatement stmt = (TSelectSqlStatement) key;
			if ( stmt.getSetOperatorType( ) == ESetOperatorType.none )
				continue;

			resultSets.add( stmt );
		}
		return resultSets;
	}

	public static List<TCTE> getCTEs( )
	{
		List<TCTE> resultSets = new ArrayList<TCTE>( );

		Iterator iter = modelBindingMap.keySet( ).iterator( );
		while ( iter.hasNext( ) )
		{
			Object key = iter.next( );
			if ( !( key instanceof TCTE ) )
			{
				continue;
			}

			TCTE cte = (TCTE) key;
			resultSets.add( cte );
		}
		return resultSets;
	}

	public static List<TTable> getTableWithSelectSetResultSets( )
	{
		List<TTable> resultSets = new ArrayList<TTable>( );

		Iterator iter = modelBindingMap.keySet( ).iterator( );
		while ( iter.hasNext( ) )
		{
			Object key = iter.next( );
			if ( !( key instanceof TTable ) )
			{
				continue;
			}

			if ( ( (TTable) key ).getSubquery( ) == null )
				continue;
			TSelectSqlStatement stmt = ( (TTable) key ).getSubquery( );
			if ( stmt.getSetOperatorType( ) == ESetOperatorType.none )
				continue;

			resultSets.add( (TTable) key );
		}
		return resultSets;
	}

	public static List<TParseTreeNode> getMergeResultSets( )
	{
		List<TParseTreeNode> resultSets = new ArrayList<TParseTreeNode>( );

		Iterator iter = modelBindingMap.keySet( ).iterator( );
		while ( iter.hasNext( ) )
		{
			Object key = iter.next( );
			if ( !( key instanceof TMergeUpdateClause )
					&& !( key instanceof TMergeInsertClause ) )
			{
				continue;
			}
			TParseTreeNode resultset = (TParseTreeNode) key;
			resultSets.add( resultset );
		}
		return resultSets;
	}

	public static List<TParseTreeNode> getUpdateResultSets( )
	{
		List<TParseTreeNode> resultSets = new ArrayList<TParseTreeNode>( );

		Iterator iter = modelBindingMap.keySet( ).iterator( );
		while ( iter.hasNext( ) )
		{
			Object key = iter.next( );
			if ( !( key instanceof TUpdateSqlStatement ) )
			{
				continue;
			}
			TParseTreeNode resultset = (TParseTreeNode) key;
			resultSets.add( resultset );
		}
		return resultSets;
	}

	public static void addRelation( Relation relation )
	{
		if ( relation != null && !relationHolder.contains( relation ) )
		{
			relationHolder.add( relation );
		}
	}

	public static Relation[] getRelations( )
	{
		return relationHolder.toArray( new Relation[0] );
	}

	public static void reset( )
	{
		relationHolder.clear( );
		modelBindingMap.clear( );
		viewModelBindingMap.clear( );
		insertModelBindingMap.clear( );
		mergeModelBindingMap.clear( );
		updateModelBindingMap.clear( );
		createModelBindingMap.clear( );
		createModelBindingMap.clear( );
		createModelQuickBindingMap.clear( );
		tableSet.clear( );
		tableAliasMap.clear( );
	}

	public static void bindCursorModel( TCursorDeclStmt stmt,
			CursorResultSet resultSet )
	{
		createModelBindingMap.put( stmt.getCursorName( ).toScript( ),
				resultSet );
	}

	public static void bindCursorIndex( TObjectName indexName,
			TObjectName cursorName )
	{
		bindModel( indexName.toScript( ),
				modelBindingMap.get( ( (CursorResultSet) createModelBindingMap
						.get( cursorName.toScript( ) ) )
								.getResultColumnObject( ) ) );
	}

}
